asynction/security.py

```diff
@@ -0,0 +1,401 @@
+SecurityCheckFactory = Callable[
+    [InternalSecurityRequirement, SecurityScheme], Optional[InternalSecurityCheck]
+]
+
+
+def unpack_security_requirement(
+    requirement: SecurityRequirement,
+) -> InternalSecurityRequirement:
+    return next(iter(requirement.items()))
+
+
+def unpack_security_requirements(
+    requirements: Sequence[SecurityRequirement],
+) -> Sequence[InternalSecurityRequirement]:
+    return list(map(unpack_security_requirement, requirements))
+
+
+def extract_auth_header(request: Request) -> Optional[Tuple[str, str]]:
+    authorization = request.headers.get("Authorization")
+
+    if not authorization:
+        return None
+    try:
+        lhs, rhs = authorization.split(None, 1)
+        return lhs, rhs
+    except ValueError as err:
+        raise SecurityException from err
+
+
+def validate_basic(
+    request: Request, basic_info_func: BasicInfoFunc, required_scopes: Sequence[str]
+) -> Optional[Mapping]:
+    auth = extract_auth_header(request)
+    if not auth:
+        return None
+    auth_type, user_pass = auth
+    if not auth_type or not user_pass:
+        raise SecurityException
+
+    if HTTPAuthenticationScheme(auth_type.lower()) != HTTPAuthenticationScheme.BASIC:
+        return None
+
+    try:
+        username, password = base64.b64decode(user_pass).decode("latin1").split(":", 1)
+    except Exception as err:
+        raise SecurityException from err
+
+    if not username or not password:
+        raise SecurityException
+
+    token_info = basic_info_func(username, password, required_scopes)
+    if token_info is None:
+        raise SecurityException
+
+    return token_info
+
+
+def validate_authorization_header(
+    request: Request, token_info_func: TokenInfoFunc
+) -> Optional[Mapping]:
+    """Check that the provided request contains a properly formatted Authorization
+    header and invokes the token_info_func on the token inside of the header.
+    """
+    auth = extract_auth_header(request)
+    if not auth:
+        return None
+    auth_type, token = auth
+    if not auth_type or not token:
+        raise SecurityException
+
+    if auth_type.lower() != "bearer":
+        return None
+
+    token_info = token_info_func(token)
+    if token_info is None:
+        raise SecurityException
+
+    return token_info
+
+
+def validate_api_key(
+    request: Request,
+    api_key_info_func: APIKeyInfoFunc,
+    required_scopes: Sequence[str],
+    bearer_format: Optional[str] = None,
+) -> Optional[Mapping]:
+    """
+    Adapted from: https://github.com/zalando/connexion/blob/main/connexion/security/security_handler_factory.py#L221  # noqa: 501
+    """
+    auth = extract_auth_header(request)
+    if not auth:
+        return None
+    auth_type, token = auth
+
+    if not auth_type or not token:
+        raise SecurityException
+
+    if auth_type.lower() != "bearer":
+        return None
+
+    token_info = api_key_info_func(token, required_scopes, bearer_format)
```

Comment:
Shouldn't the `api_key_info_func` only require 2 parameters: `api_key` (which I guess in this case is `token`, worth renaming `api_key` I guess) and `required_scopes`? Why do we also pass `bearer_format`?